### ISSUE TITLE
non static progress func callback for c++ 

### ISSUE THREAD

**Issue by kangjoni76** [curl/curl#806]
### I did this
I have c++ class that managing widget progress bar and my progressbar object is private object one. I had seen progressfunc.c example reference, but unfortunately the example expecting static function only. Thus, it was impossible to access private object in class from static function in c++.

### I expected the following
I hope I can write something like this.
 
````c++
class Downloader
{
...
 int xferinfo(void *p,
                    curl_off_t dltotal, curl_off_t dlnow,
                    curl_off_t ultotal, curl_off_t ulnow)
{
…
// here my progresswdgt call setvalue to update ui progress bar
progresswdgt->setvalue(dlnow);

…
}

private:
 widget *progresswdgt;
};
````
### curl/libcurl version
any version
[curl -V output perhaps?]

### operating system
windows and Linux, probably android

**Comment by falconindy** [comment]
tl;dr: It's simply not possible to use non-static class method as a callback.

A non-static class function is just one which has an implicit argument of "this" passed as the first argument, i.e. given:

```c++
class Foo {
 public:
  Frob(const std::string& thing);
}
```

`Frob` is essentially just a free function with a signature of `Frob(Foo* this, const std::string& thing)`. Therefore, you can't possibly use a non-static method as a callback to curl because you can't bind the correct value of `this` alongside the method. Even in C++ (remember curl is only a C library) you have to use tools like std::bind to create this association.

You should instead use a free function or static class function as a trampoline, passing along your Downloader as the CURLOPT_XFERINFODATA, e.g.

```c++
class Downloader {
 public:
  void set_progress(curl_off_t progress) { progresswdgt_->setvalue(progress); }

 private:
  widget* progresswidget_;
};

int xferinfo(void *userdata, curl_off_t dltotal, curl_off_t dlnow, curl_off_t, curl_off_t) {
  auto downloader = static_cast<Downloader*>(userdata);
  downloader->set_progress(dlnow);
}

// and later, wherever your curl options are set
curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, xferinfo);
curl_easy_setopt(curl, CURLOPT_XFERINFODATA, downloader);
```

**Comment by jay** [comment]
Yup, use [CURLOPT_XFERINFODATA](https://curl.haxx.se/libcurl/c/CURLOPT_XFERINFODATA.html) to pass a user pointer to your callback. This is an issue tracker and in the future ask questions like this on one of the [mailing lists](https://curl.haxx.se/mail/) please.
